test/model/typing_status_test.dart

```diff
@@ -0,0 +1,207 @@
+      checkTypists({dmNarrow: [eg.otherUser]});
+      checkNotifiedOnce();
+
+      model.handleTypingEvent(eg.typingEvent(groupNarrow, TypingOp.start, eg.thirdUser.userId));
+      checkTypists({dmNarrow: [eg.otherUser], groupNarrow: [eg.thirdUser]});
+      checkNotifiedOnce();
+    });
+
+    test('add a typist in the same narrow', () {
+      prepareModel();
+
+      model.handleTypingEvent(eg.typingEvent(groupNarrow, TypingOp.start, eg.otherUser.userId));
+      checkTypists({groupNarrow: [eg.otherUser]});
+      checkNotifiedOnce();
+
+      model.handleTypingEvent(eg.typingEvent(groupNarrow, TypingOp.start, eg.thirdUser.userId));
+      checkTypists({groupNarrow: [eg.otherUser, eg.thirdUser]});
+      checkNotifiedOnce();
+    });
+
+    test('sort dm recipients', () {
+      prepareModel();
+
+      final eventUnsorted = TypingEvent(id: 1, op: TypingOp.start, senderId: eg.otherUser.userId,
+        messageType: MessageType.direct,
+        recipientIds: [5, 4, 10, 8, 2, 1, eg.selfUser.userId],
+        streamId: null,
+        topic: null);
+      // DmNarrow's constructor expects the recipient IDs to be sorted.
+      check(() => model.handleTypingEvent(eventUnsorted)).returnsNormally();
+    });
+
+  });
+
+  group('handle typing stop events', () {
+    test('remove a typist from an unknown narrow', () async {
+      prepareModel();
+
+      model.handleTypingEvent(eg.typingEvent(groupNarrow, TypingOp.start, eg.otherUser.userId));
+      checkTypists({groupNarrow: [eg.otherUser]});
+      checkNotifiedOnce();
+
+      model.handleTypingEvent(eg.typingEvent(dmNarrow, TypingOp.stop, eg.otherUser.userId));
+      checkTypists({groupNarrow: [eg.otherUser]});
+      checkNotNotified();
+    });
+
+    test('remove one of two typists in the same narrow', () async {
+      prepareModel();
+
+      model.handleTypingEvent(eg.typingEvent(groupNarrow, TypingOp.start, eg.otherUser.userId));
+      checkNotifiedOnce();
+      model.handleTypingEvent(eg.typingEvent(groupNarrow, TypingOp.start, eg.thirdUser.userId));
+      checkNotifiedOnce();
+      checkTypists({groupNarrow: [eg.otherUser, eg.thirdUser]});
+
+      model.handleTypingEvent(eg.typingEvent(groupNarrow, TypingOp.stop, eg.otherUser.userId));
+      checkTypists({groupNarrow: [eg.thirdUser]});
+      checkNotifiedOnce();
+    });
+
+    test('remove typists from different narrows', () async {
+      prepareModel();
+
+      model.handleTypingEvent(eg.typingEvent(dmNarrow, TypingOp.start, eg.otherUser.userId));
+      checkNotifiedOnce();
+      model.handleTypingEvent(eg.typingEvent(groupNarrow, TypingOp.start, eg.thirdUser.userId));
+      checkNotifiedOnce();
+      model.handleTypingEvent(eg.typingEvent(topicNarrow, TypingOp.start, eg.fourthUser.userId));
+      checkNotifiedOnce();
+      checkTypists({
+          dmNarrow: [eg.otherUser],
+          groupNarrow: [eg.thirdUser],
+          topicNarrow: [eg.fourthUser]});
```

Comment:
nit: two-space indent
```suggestion
      checkTypists({
        dmNarrow: [eg.otherUser],
        groupNarrow: [eg.thirdUser],
        topicNarrow: [eg.fourthUser]});
```